src/macroxs_operations.F90

```diff
@@ -0,0 +1,243 @@
+    type is (MacroXSIso)
+      call macroxs_sample_scatter(this % scatter, gin, gout, mu, wgt)
+    type is (MacroXSAngle)
+      call find_angle(this % polar, this % azimuthal, uvw, iazi, ipol)
+      call macroxs_sample_scatter(this % scatter(iazi,ipol) % obj,gin,gout,mu,wgt)
+    end select
+
+  end subroutine sample_scatter
+
+!===============================================================================
+! MACROXS_SAMPLE_SCATTER performs the work with ScattData to sample outgoing
+! energy and change in angle.
+!===============================================================================
+
+  subroutine macroxs_sample_scatter(scatt, gin, gout, mu, wgt)
+    class(ScattData), intent(in)    :: scatt ! Scattering Object to Use
+    integer,          intent(in)    :: gin   ! Incoming neutron group
+    integer,          intent(out)   :: gout  ! Sampled outgoin group
+    real(8),          intent(out)   :: mu    ! Sampled change in angle
+    real(8),          intent(inout) :: wgt   ! Particle weight
+
+    real(8) :: xi     ! Our random number
+    real(8) :: prob   ! Running probability
+    integer :: imu
+    real(8) :: u, f, M
+    real(8) :: mu0, frac, mu1
+    real(8) :: c_k, c_k1, p0, p1
+    integer :: k, NP, samples
+
+    xi = prn()
+    prob = ZERO
+    gout = 0
+
+    do while (prob < xi)
+      gout = gout + 1
+      prob = prob + scatt % energy(gout,gin)
+    end do
+
+    select type (scatt)
+    type is (ScattDataHistogram)
+      xi = prn()
+      if (xi < scatt % data(1,gout,gin)) then
+        imu = 1
+      else
+        imu = binary_search(scatt % data(:,gout,gin), &
+                            size(scatt % data(:,gout,gin)), xi)
+      end if
+
+      ! Randomly select a mu in this bin.
+      mu = prn() * scatt % dmu + scatt % mu(imu)
+
+    type is (ScattDataTabular)
+      ! determine outgoing cosine bin
+      NP = size(scatt % data(:,gout,gin))
+      xi = prn()
+
+      c_k = scatt % data(1,gout,gin)
+      do k = 1, NP - 1
+        c_k1 = scatt % data(k+1,gout,gin)
+        if (xi < c_k1) exit
+        c_k = c_k1
+      end do
+
+      ! check to make sure k is <= NP - 1
+      k = min(k, NP - 1)
+
+      p0  = scatt % fmu(k,gout,gin)
+      mu0 = scatt % mu(k)
+      ! Linear-linear interpolation to find mu value w/in bin.
+      p1  = scatt % fmu(k+1,gout,gin)
+      mu1 = scatt % mu(k+1)
+
+      frac = (p1 - p0)/(mu1 - mu0)
+
+      if (frac == ZERO) then
+        mu = mu0 + (xi - c_k)/p0
+      else
+        mu = mu0 + (sqrt(max(ZERO, p0*p0 + TWO*frac*(xi - c_k))) - p0)/frac
+      end if
+
+      if (mu <= -ONE) then
+        mu = -ONE
+      else if (mu >= ONE) then
+        mu = ONE
+      end if
+
+    type is (ScattDataLegendre)
+      ! Now we can sample mu using the legendre representation of the scattering
+      ! kernel in data(1:this % order)
+
+      ! Do with rejection sampling
+      ! Set upper bound (instead of searching for max - though this is inefficient)
+      M = 4.0_8
```

Comment:
This also worries me because it effectively chops the PDF off at a value of 4. There is a slightly better method in the literature (if you dig deep enough) for [rejection sampling a Legendre expansion](http://www.iaea.org/inis/collection/NCLCollectionStore/_Public/07/223/7223918.pdf) (see page 10). The upper bound in Bending of ![bending upper bound](http://www.sciweavers.org/tex2img.php?eq=%5Cfrac%7B1%7D%7B%5Csqrt%7B1%20-%20%5Cmu%5E2%7D%7D&bc=White&fc=Black&im=png&fs=9&ff=arev&edit=0) can be improved to ![improved upper bound](http://www.sciweavers.org/tex2img.php?eq=%5Cfrac%7B1%7D%7B%281%20-%20%5Cmu%5E2%29%5E%7B1%2F4%7D%7D&bc=White&fc=Black&im=png&fs=9&ff=arev&edit=0) (see [this paper](http://dx.doi.org/10.1006/jath.1998.3207)). The upper bound tends to be good for low orders, but of course at high orders the rejection efficiency can get very small. That being said, it's going to be better than just drawing a rectangle around the function.

Anyway, in the interest of correctness, you can either try implementing the fancier method I mention above or explicitly find the maximum of the PDF so that you're not chopping it off.

I'm willing to give $100 to anyone who can come up with a way to directly sample an arbitrary-order Legendre expansion! I've thought about it on numerous occasions and can't come up with anything. The cdf is invertible up to order 2 and then beyond that it becomes intractable...